### PR TITLE
 Refactor state machine to prepare for methods

### PR DESCRIPTION
**Pull request by BertKleewein**
This builds on top of my previous PR. (https://github.com/Azure/azure-iot-sdk-python-preview/pull/62) 

To prepare for methods, I'm doing the following:
1. Make subscribe and unsubscribe into generic functions
2. Make those functions into "Actions" which can be queued up in the state machine once the machine is connected.  
3. Refactor the previous "send-event" code to execute on any queued actions.  This currently includes subscribe, unsubscribe, send-event.  I have a stubbed-out send-method-response action.

This is pushing the state machine into a place where there are 4 states (disconnected, connecting, connected, and disconnecting), and all operations turn into actions that go into a queue that gets processed at the appropriate time.  This is a departure from how we did it with the node sdk, but I think it can solve the same goals (for MQTT at least) with much more simplicity.  For the record, I think we'll have to add new states when we implement Twin (because we will need subscribe to be blocking), but that's a story for another time.  

This gives Carter a base to build the Method code on top of.  It still has the following weaknesses:
1. thread safety is not guaranteed.  See my previous PR-62
2. Subscribe and unsubscribe are actions just like "send event" and "send method response" and, as such, can be overlapped with any other actions.  This overlap is OK for now, but won't be forever.
3. Subscribe and unsubscribe actions go into the queue as lifo operations.  They probably need to go first, so we know that subscribes are complete before we do anything else.
4. There are transitions missing for some edge-cases (send-event while disconnecting, etc), and error handling needs to be shored up.

### CODE REVIEW

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 1]
```diff
@@ -28,6 +28,35 @@
 TOPIC_POS_INPUT_NAME = 5
 
 
+class TransportAction:
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Inherit from object (i.e. TransportAction(object)) for best compatibility

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 2]
```diff
@@ -39,13 +68,10 @@ def __init__(self, auth_provider):
         self._mqtt_provider = None
         self.on_transport_connected = None
         self.on_transport_disconnected = None
-        self.on_event_sent = None
-        self._event_queue = queue.LifoQueue()
-        self._event_callback_map = {}
+        self._action_queue = queue.Queue()
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Why the change from LIFO to FIFO? I agree that FIFO seems better, but I wonder why we were even using LIFO to begin with and what the motivation was.

Comment (BertKleewein) [reply azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Lifo was a mistake.  It should have been fifo all along.

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 3]
```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
...
+            topic = "TODO"
+            mid = self._mqtt_provider.publish(topic, action.method_response)
+            self._callback_map[mid] = action.callback
+
+        else:
+            logger.error("Removed unknown action type from queue.")
+
+    def _do_actions_in_queue(self, event_data):
         """
         Publish any events that are waiting in the event queue.  This function
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: >Publish any events that are waiting in the event queue. [](start = 8, length = 55)

Doc is outdated

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 4]
```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
             pass  # is there any other case
 
     def _on_provider_unsubscribe_complete(self, mid):
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Missing doc

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 5]
```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
...
-        logger.info("Adding event to queue for later sending")
-        # TODO: throw here if event_data.args[0] is not an event/message
-        self._event_queue.put_nowait((event_data.args[0], event_data.args[1]))
+        action = event_data.args[0]
+        if isinstance(action, TransportAction):
+            self._action_queue.put_nowait(event_data.args[0])
+        else:
+            assert False
+
+    def _do_single_action(self, action):
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: missing doc (I know its pretty self explanatory, but consistency)

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 6]
```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
...
                 return
-            logger.info("retrieved event from queue. publishing.")
-
-            encoded_topic = _encode_properties(message_to_send, self.topic)
 
-            mid = self._mqtt_provider.publish(encoded_topic, message_to_send.data)
-            self._event_callback_map[mid] = callback
+            self._do_single_action(action)
 
     def _create_mqtt_provider(self):
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Missing doc

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 7]
```diff
@@ -83,22 +109,22 @@ def __init__(self, auth_provider):
                 "dest": "disconnected",
             },
             {
-                "trigger": "_trig_send_event",
+                "trigger": "_trig_queue_action",
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Naming. I'm not clear from this what it does, exactly.

From what I can tell from the code, and our previous discussion, it seems that this enqueues an action, then potentially pops an action off the queue to resolve depending on the state.

Perhaps a better name would just be _trig_action?

**Comment by cartertinney** [comment]
>     def _on_provider_publish_complete(self, mid):

What is this parameter? It's undocumented

---
Refers to: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py:233 in c351c16. [](commit_id = c351c16c962f8163ccbfe3300a1020b29cce83f2, deletion_comment = False)

**Comment by cartertinney** [comment]
>     def _on_provider_publish_complete(self, mid):

since publish_event became do_action, are we still keeping the "publish" term around elsewhere? Seems asymmetrical.

---
Refers to: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py:233 in c351c16. [](commit_id = c351c16c962f8163ccbfe3300a1020b29cce83f2, deletion_comment = False)

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 8]
```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
...
+            self._action_queue.put_nowait(event_data.args[0])
+        else:
+            assert False
+
+    def _do_single_action(self, action):
+        if isinstance(action, SendMessageAction):
+            logger.info("running SendMessageAction")
+            message_to_send = action.message
+            base_topic = self._get_telemetry_topic()
+
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: I don't like the semantics of deriving a generic topic from a base of the telemetry topic.

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 9]
```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
...
+            if isinstance(message_to_send, Message):
+                encoded_topic = _encode_properties(message_to_send, base_topic)
+            else:
+                encoded_topic = base_topic
+                message_to_send = Message(message_to_send)
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Is this a thing we want to do in the transport, or in the client?

I've already implemented this Message wrapping in the client, but I could ditch it in favor of doing it here.

Alternatively, we could remove it here, and keep it in the client. What do you think the most appropriate place to do this is?

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 10]
```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
...
+        if isinstance(action, TransportAction):
+            self._action_queue.put_nowait(event_data.args[0])
+        else:
+            assert False
+
+    def _do_single_action(self, action):
+        if isinstance(action, SendMessageAction):
+            logger.info("running SendMessageAction")
+            message_to_send = action.message
+            base_topic = self._get_telemetry_topic()
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: >base_topic [](start = 12, length = 10)

This isn't the base topic. It's the telemetry topic.

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 11]
```diff
@@ -349,44 +385,54 @@ def _get_telemetry_topic(self):
         topic += "/messages/events/"
         return topic
```
Comment (cartertinney) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Refactor to use _get_base_topic, like how the other specific topics work

**Comment by cartertinney** [comment]
> def transport(authentication_provider):

lets make this transport_device to make clear the distinction between this and transport_module

---
Refers to: azure-iot-hub-devicesdk/tests/transport/mqtt/test_mqtt_transport.py:73 in c351c16. [](commit_id = c351c16c962f8163ccbfe3300a1020b29cce83f2, deletion_comment = False)

**Comment by cartertinney** [comment]
Overall, before I'm able to build off of this further, I'd like to see consistent documentation.

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py (hunk 1 above)
Comment (pierreca) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: >TransportAction [](start = 6, length = 15)

all actions seem to have a callback - should that be part of the TransportAction contract?

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py (hunk 2 above)
Comment (pierreca) [reply azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: sounds reasonable if the goal is to guarantee message order - is this why? 
A small comment about the design choice would protect us from asking again in the future.

---
In reply to: [266153947](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#discussion_r266153947) [](ancestors = 266153947)

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py (hunk 1 above)
Comment (pierreca) [reply azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: also, it sounds like we are mapping actions to MQTT Packets types. I wonder if the naming should reflect that.

---
In reply to: [266183861](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#discussion_r266183861) [](ancestors = 266183861)

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 12]
```diff
@@ -234,22 +234,25 @@ def _on_provider_publish_complete(self, mid):
...
-        if mid in self._event_callback_map:
-            callback = self._event_callback_map[mid]
-            del self._event_callback_map[mid]
+        if mid in self._callback_map:
+            callback = self._callback_map[mid]
+            del self._callback_map[mid]
             callback()
+        else:
+            # TODO: tests for unkonwn MID cases
+            logger.warning("PUBACK received with unknown MID: %s", str(mid))
```
Comment (pierreca) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: >PUBACK [](start = 28, length = 6)

might be a SUBACK or an UNSUBACK now as well, right?

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 13]
```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
...
+        this queue, even if they're going to be run immediately.
         """
-        logger.info("Adding event to queue for later sending")
-        # TODO: throw here if event_data.args[0] is not an event/message
-        self._event_queue.put_nowait((event_data.args[0], event_data.args[1]))
+        action = event_data.args[0]
+        if isinstance(action, TransportAction):
+            self._action_queue.put_nowait(event_data.args[0])
+        else:
+            assert False
```
Comment (pierreca) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: >assert False [](start = 12, length = 12)

I'm not sure I understand how we could end up here. if the goal is to provoke a crash that will encourage a customer to publish a bug report, i think we should add a debug statement for that "Oops! looks like you found a bug - please submit a github issue <...instructions...>"

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py (hunk 5 above)
Comment (pierreca) [reply azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: quite helpful to every potential future contributor. let's not block on this but let's keep track of this. thanks for pushing for this Carter!

---
In reply to: [266157401](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#discussion_r266157401) [](ancestors = 266157401)

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 14]
```diff
@@ -349,44 +385,54 @@ def _get_telemetry_topic(self):
...
+    def _get_base_topic(self):
+        if self._auth_provider.module_id:
+            return (
+                "devices/"
+                + self._auth_provider.device_id
+                + "/modules/"
+                + self._auth_provider.module_id
+            )
+        else:
+            return "devices/" + self._auth_provider.device_id
```
Comment (pierreca) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: nit: this feels a little convoluted/repetitive. how about:
```
base_topic = "devices/" + self._auth_provider.device_id
if self._auth_provider.module_id:
    base_topic += "/modules/" + self._auth_provider.module_id 
```

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 15]
```diff
@@ -349,44 +385,54 @@ def _get_telemetry_topic(self):
...
 
     def disconnect(self, callback=None):
+        logger.info("disconnect called")
         self._disconnect_callback = callback
         self._trig_disconnect()
 
     def send_event(self, message, callback=None):
-        self._trig_send_event(message, callback)
+        action = SendMessageAction(message, callback)
+        self._trig_queue_action(action, self._action_queue)
```
Comment (pierreca) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: >self._action_queue [](start = 40, length = 18)

here you have a second parameter but not below. leftovers? i don't see code that changes based on a second parameter?

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py [hunk 16]
```diff
@@ -39,13 +68,10 @@ def __init__(self, auth_provider):
         self._mqtt_provider = None
         self.on_transport_connected = None
         self.on_transport_disconnected = None
-        self.on_event_sent = None
-        self._event_queue = queue.LifoQueue()
-        self._event_callback_map = {}
+        self._action_queue = queue.Queue()
+        self._callback_map = {}
```
Comment (pierreca) [inline azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: If i understand correctly, _action_queue is for "pending" actions (ie not on the wire yet) and _callback_map is to store the callbacks of actions "in flight" (ie the packet has been sent and the ACK hasn't been received). I wonder:
- if the naming could convey this idea (provided that my understanding is correct)
- (less important) if we need one additional "action processor" data structure that groups together the pending action queue, the callback map, and the methods that act on those (_add_to_queue, _do_single_action and _do_actions_in_queue). would it make this easier to test?

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py (hunk 1 above)
Comment (BertKleewein) [reply azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: added callback.
Also, this is not a 1:1 mapping to packet types.  This is a higher level.  MethodResponse and SendMessage both result in PUBLISH operations.

---
In reply to: [266184101](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#discussion_r266184101) [](ancestors = 266184101,266183861)

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py (hunk 16 above)
Comment (BertKleewein) [reply azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Awesome!  I totally agree on both points, but I want to refactor the "action processor" in a separate commit.

---
In reply to: [266185036](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#discussion_r266185036) [](ancestors = 266185036)

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py (hunk 13 above)
Comment (cartertinney) [reply azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: Adding onto this, why don't we raise an Exception here instead? After all, "assert False" raises an AssertionError, which isn't terribly informative (and is intended largely for testing) - we might as well choose what gets raised.

---
In reply to: [266184228](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#discussion_r266184228) [](ancestors = 266184228)

File: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py (hunk 9 above)
Comment (BertKleewein) [reply azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py]: I think the client is the right place.  It has to happen for all transports.  I removed it from here.

---
In reply to: [266169151](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#discussion_r266169151) [](ancestors = 266169151)

**Comment by BertKleewein** [comment]
>     def _on_provider_publish_complete(self, mid):

yes, "PUBLISH" is still an MQTT operation, so we do need to keep it around, but maybe in a more limited context.  Many of our actions will involve sending a PUBLISH to the hub.

---
In reply to: [473463623](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#issuecomment-473463623) [](ancestors = 473463623)

---
Refers to: azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py:233 in ac29e41. [](commit_id = ac29e41a97938e7fcd1192de7e581e1c571aaeae, deletion_comment = False)

**Comment by BertKleewein** [comment]
> def transport(authentication_provider):

I like that idea, but that is more test churn than I want to deal with right now.  Next time we're poking around in here, we should change it.

---
In reply to: [473467073](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#issuecomment-473467073) [](ancestors = 473467073)

---
Refers to: azure-iot-hub-devicesdk/tests/transport/mqtt/test_mqtt_transport.py:73 in ac29e41. [](commit_id = ac29e41a97938e7fcd1192de7e581e1c571aaeae, deletion_comment = False)
